lib/utils/ignore.ts

```diff
@@ -0,0 +1,151 @@
+		this.dockerIgnoreEntries = [];
+		this.gitIgnoreEntries = [];
+	}
+	/**
+	 * @param  {string} relativePath
+	 * 	The relative pathname from the build context, for example a root level .gitignore should be
+	 * 		./.gitignore
+	 * @returns IgnoreFileType
+	 * 	The type of ignore file, or null
+	 */
+	public isIgnoreFile(relativePath: string): IgnoreFileType | null {
+		for (const { pattern, type, allowSubdirs } of FileIgnorer.ignoreFiles) {
+			if (relativePath === pattern || relativePath === `./${pattern}`) {
+				return type;
+			}
+			const basename = path.posix.basename(relativePath);
+			if (allowSubdirs && basename === pattern) {
+				return type;
+			}
+		}
+
+		return null;
+	}
+	/**
+	 * @param  {string} fullPath
+	 * 	The full path on disk of the ignore file
+	 * @param  {IgnoreFileType} type
+	 * @returns Promise
+	 */
+	public async addIgnoreFile(
+		fullPath: string,
+		type: IgnoreFileType,
+	): Promise<void> {
+		const contents = await fs.readFile(fullPath, 'utf8');
+
+		contents.split('\n').forEach(line => {
+			const trimmed = line.trimLeft();
+			// ignore empty lines and comments
+			if (_.isEmpty(trimmed) || trimmed[0] === '#') {
+				return;
+			}
+			this.addEntry(trimmed, fullPath, type);
+		});
+
+		return;
+	}
+
+	// Pass this function as a predicate to a filter function, and it will filter
+	// any ignored files
+	public filter(filename: string): boolean {
+		const dockerIgnoreHandle = dockerIgnore();
+		const gitIgnoreHandle = ignore();
+
+		const relFile = path.relative(this.basePath, filename);
+
+		_.each(this.dockerIgnoreEntries, ({ pattern, filePath }) => {
+			// If this file falls under the scope of an ignore file
+			if (FileIgnorer.contains(path.posix.dirname(filePath), filename)) {
+				dockerIgnoreHandle.add(pattern);
+			}
+		});
+
+		if (dockerIgnoreHandle.ignores(relFile)) {
+			return false;
+		}
+
+		_.each(this.gitIgnoreEntries, ({ pattern, filePath }) => {
+			if (FileIgnorer.contains(path.posix.dirname(filePath), filename)) {
+				gitIgnoreHandle.add(pattern);
+			}
+		});
+
+		if (gitIgnoreHandle.ignores(relFile)) {
+			return false;
+		}
+
+		return true;
+	}
+
+	private addEntry(
+		pattern: string,
+		filePath: string,
+		type: IgnoreFileType,
+	): void {
+		const entry: IgnoreEntry = { pattern, filePath };
+		switch (type) {
+			case IgnoreFileType.DockerIgnore:
+				this.dockerIgnoreEntries.push(entry);
+				break;
+			case IgnoreFileType.GitIgnore:
+				this.gitIgnoreEntries.push(entry);
+				break;
+		}
+	}
+
+	/**
+	 * Given two paths, check whether the first contains the second
+	 * @param path1 The potentially containing path
+	 * @param path2 The potentially contained path
+	 * @return A boolean indicating whether `path1` contains `path2`
+	 */
+	private static contains(path1: string, path2: string): boolean {
+		// First normalise the input, to remove any path weirdness
+		path1 = path.posix.normalize(path1);
+		path2 = path.posix.normalize(path2);
```

Comment:
How does this behave with symlinks? Are they expanded to their target path?

Reply:
They aren't, it doesn't care about the contents, only the path. The symlinks are then handled by tar-stream (or rather not handled, and just added as symlinks).